File: Sources/KnowFlick/Services/AIService.swift
import Foundation

/// 与 OpenAI 兼容的 Chat Completions API 客户端
/// 默认对接 DeepSeek，可在设置里改 baseURL / model
struct AIService {
    // MARK: - 生成知识卡片

    struct AICardPayload: Decodable {
        let category: String
        let headline: String
        let summary: String
        let details: String
        let searchKeywords: [String]   // 用于构造科普链接
    }

    /// 请求 AI 生成 count 张知识卡，排除已有标题避免重复
    func generateCards(
        settings: AISettings,
        count: Int,
        excludeHeadlines: [String]
    ) async throws -> [KnowledgeCard] {
        guard !settings.apiKey.isEmpty else {
            throw AIError.missingKey
        }
        guard !settings.baseURL.isEmpty else {
            throw AIError.badRequest("baseURL 为空")
        }

        let systemPrompt = """
        你是一个知识卡片编辑。生成有趣、准确、有深度的冷知识卡片，面向中文读者。
        要求：
        - 每条知识必须真实可查证，严禁编造事实、数字、人名
        - 标题一句话点出反直觉或有趣的点，如「香蕉是浆果，草莓不是」
        - 摘要一句话概括核心
        - 详情 3-6 段，每段讲一个角度（机制解释、历史背景、冷门细节、相关现象），用 \\n\\n 分段
        - 给出 2-4 个建议搜索关键词，方便读者深入查证（不用给链接，给关键词即可）
        - 分类用中文短词：物理、生物、历史、数学、化学、天文、心理、语言、科技、生活、地理、脑科学等
        - 每次返回 count 条，避免与已有标题重复
        """

        let userPrompt = """
        请生成 \(count) 条不重复的冷知识卡片，严格按 JSON 数组格式输出，不要输出任何其他文字：
        [
          {
            "category": "分类",
            "headline": "标题",
            "summary": "一句话摘要",
            "details": "详情（多段，用 \\n\\n 分隔）",
            "searchKeywords": ["关键词1", "关键词2", "关键词3"]
          }
        ]

        已存在标题（避免重复）：
        \(excludeHeadlines.prefix(100).joined(separator: "\n"))

        若偏好某分类请优先该分类：\(settings.categoryFilter.isEmpty ? "不限" : settings.categoryFilter)
        """

        let messages: [[String: String]] = [
            ["role": "system", "content": systemPrompt],
            ["role": "user", "content": userPrompt]
        ]

        let raw = try await chatCompletion(messages: messages, settings: settings)

        // 解析 JSON（AI 可能包 ```json 代码块，先清理）
        let cleaned = cleanJSON(from: raw)
        guard let data = cleaned.data(using: .utf8) else { throw AIError.parse("无法编码响应") }
        let payloads: [AICardPayload]
        do {
            payloads = try JSONDecoder().decode([AICardPayload].self, from: data)
        } catch {
            // 兜底：尝试解析单个对象
            let single = try JSONDecoder().decode(AICardPayload.self, from: data)
            payloads = [single]
        }

        let now = Date()
        return payloads.map { p in
            KnowledgeCard(
                category: p.category,
                headline: p.headline,
                summary: p.summary,
                details: p.details,
                links: Self.buildSearchLinks(keywords: p.searchKeywords),
                source: .ai,
                createdAt: now
            )
        }
    }

    // MARK: - 基础调用

    private func chatCompletion(messages: [[String: String]], settings: AISettings) async throws -> String {
        var urlString = settings.baseURL
            .trimmingCharacters(in: CharacterSet(charactersIn: "/"))
        if urlString.hasSuffix("/v1") {
            urlString += "/chat/completions"
        } else {
            urlString += "/v1/chat/completions"
        }
        guard let url = URL(string: urlString) else { throw AIError.badRequest("URL 无效: \(urlString)") }

        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.timeoutInterval = 60
        request.setValue("Bearer \(settings.apiKey)", forHTTPHeaderField: "Authorization")
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")

        let body: [String: Any] = [
            "model": settings.model,
            "messages": messages,
            "temperature": 1.0,
            "max_tokens": 2048,
            "stream": false
        ]
        request.httpBody = try JSONSerialization.data(withJSONObject: body)

        let (data, response) = try await URLSession.shared.data(for: request)
        guard let http = response as? HTTPURLResponse else { throw AIError.network("无响应") }

        guard (200..<300).contains(http.statusCode) else {
            let msg = String(data: data, encoding: .utf8) ?? "HTTP \(http.statusCode)"
            throw AIError.httpStatus(http.statusCode, msg)
        }

        struct ChatResponse: Decodable {
            struct Choice: Decodable {
                struct Message: Decodable { let content: String }
                let message: Message
            }
            let choices: [Choice]
        }

        do {
            let parsed = try JSONDecoder().decode(ChatResponse.self, from: data)
            return parsed.choices.first?.message.content ?? ""
        } catch {
            throw AIError.parse("响应解析失败")
        }
    }

    // MARK: - 工具

    /// 把关键词构造成真实可用的科普检索链接（避免幻觉 URL）
    static func buildSearchLinks(keywords: [String]) -> [ScienceLink] {
        keywords.prefix(3).map { kw in
            let encoded = kw.addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed) ?? kw
            return ScienceLink(
                title: "搜索：\(kw)",
                url: "https://www.bing.com/search?q=\(encoded)"
            )
        }
    }

    /// 清理 AI 输出里的 ```json 代码块包裹
    private func cleanJSON(from raw: String) -> String {
        var text = raw.trimmingCharacters(in: .whitespacesAndNewlines)
        if text.hasPrefix("```") {
            text = text.replacingOccurrences(of: "^```(?:json)?\\s*", with: "", options: .regularExpression)
            text = text.replacingOccurrences(of: "\\s*```$", with: "", options: .regularExpression)
        }
        return text
    }
}

enum AIError: LocalizedError {
    case missingKey
    case badRequest(String)
    case network(String)
    case httpStatus(Int, String)
    case parse(String)

    var errorDescription: String? {
        switch self {
        case .missingKey: "未配置 API Key，请到设置里填写"
        case .badRequest(let s): "请求参数错误：\(s)"
        case .network(let s): "网络错误：\(s)"
        case .httpStatus(let c, let m): "API 返回 \(c)：\(String(m.prefix(200)))"
        case .parse(let s): "解析失败：\(s)"
        }
    }
}

File: android/app/src/main/kotlin/com/knowflick/app/ai/AiService.kt
package com.knowflick.app.ai

import com.knowflick.app.domain.CardSource
import com.knowflick.app.domain.KnowledgeCard
import com.knowflick.app.domain.ScienceLink
import java.net.URL
import java.net.URLEncoder
import java.util.concurrent.TimeUnit
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.delay
import kotlinx.coroutines.withContext
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.add
import kotlinx.serialization.json.buildJsonArray
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.put
import okhttp3.MediaType.Companion.toMediaTypeOrNull
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody

/**
 * OpenAI 兼容 Chat Completions 客户端（移植自 macOS AIService）：
 * - endpoint 拼接口径一致：保留服务商自定义前缀，仅裸域名补 /v1
 * - 流式生成 + 增量对象扫描，够数即停；429/5xx 指数退避重试
 * - 近重复抑制（归一化标题 + bigram Jaccard）与 ≥80 字门槛；超 6 张自动分批
 */
class AiService(
    private val client: OkHttpClient = defaultClient(),
    private val retryBaseDelayMs: Long = 1_000L,
) {
    private val json = Json { ignoreUnknownKeys = true }
    private val jsonMedia = "application/json; charset=utf-8".toMediaTypeOrNull()!!

    suspend fun ping(settings: AiSettings, apiKey: String) {
        val key = apiKey.trim()
        if (requiresKey(settings.baseURL) && key.isEmpty()) throw AiError.MissingKey()
        val body = buildJsonObject {
            put("model", settings.model.ifBlank { "gpt-4o-mini" })
            put("messages", buildJsonArray {
                add(buildJsonObject { put("role", "user"); put("content", "ping") })
            })
            put("max_tokens", 1)
            put("stream", false)
        }
        val request = Request.Builder()
            .url(completionURL(settings.baseURL))
            .post(body.toString().toRequestBody(jsonMedia))
            .header("Content-Type", "application/json")
            .header("User-Agent", userAgent())
            .apply { if (key.isNotEmpty()) header("Authorization", "Bearer $key") }
            .build()
        withContext(Dispatchers.IO) {
            client.newCall(request).execute().use { response ->
                if (!response.isSuccessful) {
                    throw AiError.HttpStatus(response.code, errorMessage(response.body?.string().orEmpty()))
                }
            }
        }
    }

    /** 生成知识卡片。onDelta 提供增量文本（生成中展示用）。上限 20；单请求 ≤6 张自动分批。 */
    suspend fun generateCards(
        settings: AiSettings,
        apiKey: String,
        count: Int,
        excludeHeadlines: List<String>,
        topic: String? = null,
        preferredSources: List<String> = emptyList(),
        onDelta: ((String) -> Unit)? = null,
    ): List<KnowledgeCard> {
        val key = apiKey.trim()
        if (requiresKey(settings.baseURL) && key.isEmpty()) throw AiError.MissingKey()
        if (settings.baseURL.isBlank()) throw AiError.BadRequest("baseURL 为空")
        if (count !in 1..20) throw AiError.BadRequest("生成数量须为 1–20")

        val seenHeadlines = excludeHeadlines.map { AiTextUtils.normalizeHeadline(it) }.toMutableSet()
        val excludedBigrams = excludeHeadlines.map { AiTextUtils.bigramSet(it) }.toMutableList()
        val excludeList = excludeHeadlines.take(100).toMutableList()
        val allCards = ArrayList<KnowledgeCard>(count)
        val now = System.currentTimeMillis()

        var remaining = count
        while (remaining > 0) {
            val batch = minOf(remaining, MAX_CARDS_PER_REQUEST)
            val payloads = requestBatch(settings, key, batch, topic, excludeList.take(100), onDelta)
            for (payload in payloads) {
                if (payload.details.length < 80) continue
                val headlineKey = AiTextUtils.normalizeHeadline(payload.headline)
                val bigram = AiTextUtils.bigramSet(payload.headline)
                if (headlineKey.isEmpty() || !seenHeadlines.add(headlineKey)) continue
                if (excludedBigrams.any { AiTextUtils.jaccard(bigram, it) > AiTextUtils.NEAR_DUPLICATE_THRESHOLD }) continue
                seenHeadlines += headlineKey
                excludedBigrams += bigram
                excludeList += payload.headline
                allCards += KnowledgeCard(
                    id = newId(),
                    category = com.knowflick.app.domain.CategoryRegistry.normalize(payload.category, custom = emptyList()),
                    headline = payload.headline,
                    summary = payload.summary,
                    details = payload.details,
                    links = AiTextUtils.buildSearchLinks(
                        keywords = payload.searchKeywords,
                        preferred = preferredSources,
                        aiSources = payload.sources,
                    ),
                    source = CardSource.AI,
                    createdAt = now,
                )
            }
            remaining -= batch
        }
        if (allCards.isEmpty()) throw AiError.NoUsableCards()
        return allCards
    }

    private suspend fun requestBatch(
        settings: AiSettings,
        apiKey: String,
        batchCount: Int,
        topic: String?,
        excludeList: List<String>,
        onDelta: ((String) -> Unit)?,
    ): List<AiCardPayload> {
        val body = buildJsonObject {
            put("model", settings.model)
            put("messages", buildJsonArray {
                add(buildJsonObject {
                    put("role", "system")
                    put("content", AiTextUtils.cardSystemPrompt(
                        categoryWhitelist = com.knowflick.app.domain.CategoryRegistry.BUILTIN_CATEGORY,
                        sourcesHint = "维基百科、国家地理、NASA 等权威科普站点",
                    ))
                })
                add(buildJsonObject {
                    put("role", "user")
                    put("content", AiTextUtils.cardUserPrompt(batchCount, topic, excludeList, categoryFilter = ""))
                })
            })
            put("temperature", 0.8)
            put("max_tokens", minOf(8192, maxOf(1200, batchCount * 900 + 400)))
            put("stream", true)
        }
        val request = Request.Builder()
            .url(completionURL(settings.baseURL))
            .post(body.toString().toRequestBody(jsonMedia))
            .header("Content-Type", "application/json")
            .header("User-Agent", userAgent())
            .apply { if (apiKey.isNotEmpty()) header("Authorization", "Bearer $apiKey") }
            .build()

        return withContext(Dispatchers.IO) {
            for (attempt in 0 until 3) {
                try {
                    client.newCall(request).execute().use { response ->
                        if (!response.isSuccessful) throw AiError.HttpStatus(response.code, errorMessage(response.body?.string().orEmpty()))
                        val source = response.body?.source() ?: throw AiError.Network("无响应体")
                        val scanner = AiObjectScanner()
                        while (true) {
                            val line = source.readUtf8Line() ?: break
                            if (line.trim() == "data: [DONE]") break
                            val delta = sseContentDelta(line)
                            if (delta.isNullOrEmpty()) continue
                            scanner.append(delta)
                            onDelta?.invoke(delta)
                            if (scanner.objects.size >= batchCount) break
                        }
                        val found = AiPayloadParser.parseObjects(scanner.objects.joinToString("") { it.toString() })
                        if (found.isEmpty()) throw AiError.Parse("AI 未返回可用回复，请重试")
                        return@withContext found
                    }
                } catch (e: AiError.HttpStatus) {
                    if ((e.code == 429 || e.code in 500..599) && attempt < 2) {
                        kotlinx.coroutines.delay(retryBaseDelayMs * (1L shl attempt))
                        continue
                    }
                    throw e
                }
            }
            throw AiError.Network("重试耗尽")
        }
    }

    companion object {
        private val jsonMedia = "application/json; charset=utf-8".toMediaTypeOrNull()
        private const val MAX_CARDS_PER_REQUEST = 6
        const val ANDROID_VERSION_NAME = "0.3.0"

        fun userAgent(): String = "KnowFlick-Android/$ANDROID_VERSION_NAME"

        /** 保留服务商自定义 API 前缀，只有裸域名才补 /v1（与 macOS 口径一致） */
        fun completionURL(baseURL: String): URL {
            val raw = baseURL.trim()
            val url = runCatching { URL(raw) }.getOrNull()
                ?: throw AiError.BadRequest("请输入有效的 HTTP(S) API Base URL")
            if (url.protocol !in listOf("https", "http")) throw AiError.BadRequest("请输入有效的 HTTP(S) API Base URL")
            var path = url.path ?: ""
            while (path.endsWith("/")) path = path.dropLast(1)
            if (!path.endsWith("/chat/completions")) {
                path += if (path.isEmpty()) "/v1/chat/completions" else "/chat/completions"
            }
            return URL(url.protocol, url.host, url.port, path)
        }

        /** "data: {json}" → 内容增量（兼容 delta.content / message.content） */
        fun sseContentDelta(line: String): String? {
            val trimmed = line.trim()
            if (!trimmed.startsWith("data:")) return null
            val payload = trimmed.drop(5).trim()
            if (payload.isEmpty() || payload == "[DONE]") return null
            return runCatching {
                val obj = Json.parseToJsonElement(payload).jsonObjectSafe()
                val choices = obj["choices"] as? kotlinx.serialization.json.JsonArray ?: return null
                val first = choices.firstOrNull() as? kotlinx.serialization.json.JsonObject ?: return null
                val delta = first["delta"] as? kotlinx.serialization.json.JsonObject
                val message = first["message"] as? kotlinx.serialization.json.JsonObject
                delta?.get("content")?.primitiveContent()
                    ?: message?.get("content")?.primitiveContent()
            }.getOrNull()
        }

        /** 从错误响应体提取 message 字段（上限 600 字符） */
        fun errorMessage(body: String): String = runCatching {
            val obj = Json.parseToJsonElement(body).jsonObjectSafe()
            val error = obj["error"] as? kotlinx.serialization.json.JsonObject
            (error?.get("message") as? kotlinx.serialization.json.JsonPrimitive)?.content
                ?: (obj["message"] as? kotlinx.serialization.json.JsonPrimitive)?.content
        }.getOrNull() ?: body.take(600)

        /** 本地部署识别：回环主机或 Ollama 默认端口（与 macOS 表驱动口径一致） */
        fun requiresKey(baseURL: String): Boolean = !AiProviderPresets.match(baseURL).let { it.id in AiProviderPresets.keylessIds }

        private fun defaultClient(): OkHttpClient = OkHttpClient.Builder()
            .connectTimeout(30, TimeUnit.SECONDS)
            .readTimeout(180, TimeUnit.SECONDS)
            .writeTimeout(30, TimeUnit.SECONDS)
            .build()

    }
}

private fun newId(): String = java.util.UUID.randomUUID().toString().uppercase()

private fun kotlinx.serialization.json.JsonElement.jsonObjectSafe(): kotlinx.serialization.json.JsonObject =
    this as kotlinx.serialization.json.JsonObject

private fun kotlinx.serialization.json.JsonElement.primitiveContent(): String? =
    (this as? kotlinx.serialization.json.JsonPrimitive)?.content

